specs/core/0_beacon-chain.md

```diff
@@ -570,6 +574,8 @@ The following data structures are defined as [SimpleSerialize (SSZ)](https://git
     'initiated_exit': 'bool',
     # Was the validator slashed
     'slashed': 'bool',
+    # Rounded balance
+    'high_balance': 'uint32'
```

Comment:
Can we make `high_balance` of type `Gwei`? Benefits:

* It avoids introducing a new semantic type (`ETH`).
* It simplifies `get_balance` and `set_balance` (see [here](https://github.com/ethereum/eth2.0-specs/issues/685#issuecomment-467025467)).
* It is slightly more future-proof.

While the 2x performance benefit for `hash_tree_root(state.low_balances)` (saving 32MB of hashing per epoch in the worst case) is nice, I'm also tempted to make `low_balances` of type `Gwei` for simplicity:

* It avoids introducing a new SSZ type (`uint32`).
* It avoids introducing a new semantic type (`Uint32Gwei`).
* We still get the significant light client optimisation.

Reply:
The benefits of halving the validator registry recomputation time are large. If we can derive from protolambda's shuffling statistics that it takes 1.25s to hash 45 bits * 4 million validators, then cutting low balances to 32 bits would cut the maximum validator balance re-hashing time from 1.25 * 128/45 = 3.55 seconds to 1.25 * 64/45 = 1.78 seconds, which is potentially the difference between the chain having a hiccup every six minutes and not.

> It avoids introducing a new SSZ type (uint32).

SSZ already has a `unit32` type, and it's a type we'll need anyway if/when we end up [doing CBC](https://github.com/ethereum/eth2.0-specs/issues/701), for example.

I'd be ok with the high balance being a uint64. But note that it should not be of type Gwei because it's not counting Gwei, it's counting multiples of a billion Gwei. This is the same reasoning as why you would not use the Epoch type to refer to, say, proof of custody periods. Perhaps we should just call the types of the variables in the state uint32 and/or uint64, and use the Gwei type only for numbers that get passed into or out of `get_balance` and `set_balance` calls.

Reply:
> cutting low balances to 32 bits would cut the maximum validator balance re-hashing time from 1.25 * 128/45 = 3.55 seconds to 1.25 * 64/45 = 1.78 seconds

Shuffling is somewhat different than Merkleisation. I just did a quick benchmark in Go where `hash_tree_root(state.low_balances)` takes 544ms for 2^22 validators when using `uint64`. So going down to `uint32` would at most shave 272ms per epoch. (My benchmark is single-thread and not necessarily optimised.)

> it's not counting Gwei

Right. My suggestion is to do the accounting in Gwei (not multiples of a billion Gwei). This simplifies both the typing and the `get_balance `/`set_balance` code.

Reply:
Just to provide a bit more context here, shuffling does 37-byte input hashes which take 300ns and merkleisation does 64-byte input hashes which take 500ns.

Another consideration is that in the current 1.25s shuffling the hash function consumes at most 25% of the total time. We discussed a possible optimisation with @protolambda to improve the 1.25s number.

Reply:
> We discussed a possible optimisation with @protolambda to improve the 1.25s number.

Have you done the optimization that requires you to only hash half the whole range of indices every round already? (This is the one that allows shuffling to be computed with 45 bits of hashing per validator instead of 90)

BTW I ran it in python and you are right, going down from 64 to 32 shaves only 350ms (in python as usual, hashing is artificially negligible relative to everything else because it's one of the few things that's C-optimized; the actual reduction was from 4150 ms to 3700).

Would it satisfy your uint64 purism bone to have the value be specified as bytes4 instead of uint32? I see the "low balance" as not being a real standalone "number" but rather a backend thing whose role is abstracted away by the `get_balance` and `set_balance` functions.

Reply:
> Have you done the optimization that requires you to only hash half the whole range of indices every round already?

I believe that optimisation was done from the start. More advanced optimisations are being studied. We're somewhat hopeful we can get shuffling of 4,000,000 validators closer to 0.5s. cc @protolambda

> going down from 64 to 32 shaves only 350ms

One of the things we are looking into is drastically reducing that number. Merkleisation is extremely parallelisable so the data to be hashed can be pipelined to the CPU to max out the bandwidth of the SHA256 instructions. (Further gains can be had with things like AVX2 and AVX512.)

One good reason to optimise data structures for hashing is in preparation for a STARK-friendly hash function, so I'm definitely weakening my stance that this is a micro-optimisation.

> Would it satisfy your uint64 purism bone to have the value be specified as bytes4 instead of uint32?

Interesting! Unfortunately the Merkleisation spec only does packing on basic types (`byte[4]` is a tuple). My purism bone is merely a "strong opinion weakly held" to guide intuition. It should be jettisoned if necessary.

Do we at least agree that `high_balance` should be `Gwei`? 😄

Reply:
Sure, I'm fine with `high_balance` being Gwei. If we do end up not wanting to use a uint32 for low balance, then I think we should simplify and just store the entire balance in the low balance field, so balance reads are unchanged and it's only writes that have an extra update step.